### PR TITLE
pantheon.switchboard-plug-parental-controls: init at 8.0.1

### LINKED ISSUE #138422: Package Switchboard Screen Time & Limits Plug
**Project description**
Settings for Pantheon desktop.

This might help me to limit my screen time and actually sleep at night instead of scrolling through social media.

I'm not sure how easily it would integrate into nixos.

cc @bobby285271 

**Metadata**

* homepage URL: https://github.com/elementary/switchboard-plug-parental-controls
* source URL: https://github.com/elementary/switchboard-plug-parental-controls
* license: GPLv3+ ([source](https://github.com/elementary/switchboard-plug-parental-controls/blob/master/src/plug/Plug.vala#L8))
* platforms: linux

### PR DESCRIPTION
**Pull request by davidak**
###### Motivation for this change

Close #138422 

Screenshot from 2021-10-04 08-04-50

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by davidak** [comment]
It currently fails to build.

@bobby285271 have you seen this issue before?

```
[davidak@gaming:~/code/nixpkgs]$ nix-build . -A pantheon.switchboard-plug-parental-controls
these derivations will be built:
  /nix/store/dvqi8cdzc21gg7qzjz1ccildgg4fjky2-switchboard-plug-parental-controls-6.0.0.drv
building '/nix/store/dvqi8cdzc21gg7qzjz1ccildgg4fjky2-switchboard-plug-parental-controls-6.0.0.drv'...
unpacking sources
unpacking source archive /nix/store/xwfd49jzdr7z3490zs6algqhln7cdxpa-source
source root is source
patching sources
configuring
meson flags: --buildtype=plain         --libdir=/nix/store/xz0lxir7iglzcw36zmcw3i6ldysbaz1x-switchboard-plug-parental-controls-6.0.0/lib --libexecdir=/nix/store/xz0lxir7iglzcw36zmcw3i6ldysbaz1x-switchboard-plug-parental-controls-6.0.0/libexec         --bindir=/nix/store/xz0lxir7iglzcw36zmcw3i6ldysbaz1x-switchboard-plug-parental-controls-6.0.0/bin --sbindir=/nix/store/xz0lxir7iglzcw36zmcw3i6ldysbaz1x-switchboard-plug-parental-controls-6.0.0/sbin         --includedir=/nix/store/xz0lxir7iglzcw36zmcw3i6ldysbaz1x-switchboard-plug-parental-controls-6.0.0/include         --mandir=/nix/store/xz0lxir7iglzcw36zmcw3i6ldysbaz1x-switchboard-plug-parental-controls-6.0.0/share/man --infodir=/nix/store/xz0lxir7iglzcw36zmcw3i6ldysbaz1x-switchboard-plug-parental-controls-6.0.0/share/info         --localedir=/nix/store/xz0lxir7iglzcw36zmcw3i6ldysbaz1x-switchboard-plug-parental-controls-6.0.0/share/locale         -Dauto_features=enabled         -Dwrap_mode=nodownload         --prefix=/nix/store/xz0lxir7iglzcw36zmcw3i6ldysbaz1x-switchboard-plug-parental-controls-6.0.0
The Meson build system
Version: 0.57.1
Source dir: /build/source
Build dir: /build/source/build
Build type: native build
Project name: parental-controls
Project version: 6.0.0
C compiler for the host machine: gcc (gcc 10.3.0 "gcc (GCC) 10.3.0")
C linker for the host machine: gcc ld.bfd 2.35.1
Vala compiler for the host machine: valac (valac 0.52.5)
Host machine cpu family: x86_64
Host machine cpu: x86_64
Found pkg-config: /nix/store/abzvpq14cipsiv3nn7rx800na8ig1j3f-pkg-config-wrapper-0.29.2/bin/pkg-config (0.29.2)
Run-time dependency polkit-gobject-1 found: YES 0.119
Run-time dependency glib-2.0 found: YES 2.68.4
Run-time dependency gio-2.0 found: YES 2.68.4
Run-time dependency gio-unix-2.0 found: YES 2.68.4
Run-time dependency gobject-2.0 found: YES 2.68.4
Run-time dependency granite found: YES 6.1.1
Run-time dependency gtk+-3.0 found: YES 3.24.30
Run-time dependency gee-0.8 found: YES 0.20.4
Run-time dependency flatpak found: YES 1.10.2
Run-time dependency malcontent-0 found: YES 0.10.1
Run-time dependency libhandy-1 found: YES 1.4.0
Run-time dependency accountsservice found: YES 0.6.55
Run-time dependency systemd found: YES 249
Library posix found: YES
Run-time dependency dbus-1 found: YES 1.12.20
Configuring pantheon-parental-controls.service using configuration
Configuring screentime-limits.policy.in using configuration
Configuring org.pantheon.ParentalControls.service using configuration
Configuring pantheon-parental-controls-client.desktop using configuration
Configuring Constants.vala using configuration
Run-time dependency switchboard-2.0 found: YES 6.0.0
Build targets in project: 11

Found ninja-1.10.2 at /nix/store/dvdph40kdis6mgdi47w57pkmr00k5n5w-ninja-1.10.2/bin/ninja
meson: enabled parallel building
building
build flags: -j16 -l16
[17/44] Compiling Vala source ../src/plug/MainView.vala ../src/plug/Plug.vala ../src/plug/Views/AppsView.vala ../src/plug/Views/TimeLimitView.vala ../src/plug/Views/InternetView.vala ../src/plug/Widgets/UserList.vala ../src/plug/Widgets/UserItem.vala ../src/plug/Widgets/ControlPage.vala ../src/plug/Widgets/AppChooser.vala ../src/plug/Widgets/AppRow.vala ../src/common/Utils.vala src/common/Constants.vala ../src/common/PAM/Writer.vala ../src/common/PAM/Token.vala ../src/common/PAM/Reader.valaonstants.vala ../src/common/PAM/Writer.vala ../src/common/PAM/Token.vala ../src/common/PAM/Reader.vala
../src/plug/Widgets/UserItem.vala:98.13-98.38: warning: `Hdy.Avatar.set_image_load_func' has been deprecated since 1.2
Compilation succeeded - 1 warning(s)
[44/44] Linking target src/plug/libparental-controls.soo.p/meson-generated_Utils.c.oUtils.c.oKKDialog.c.on.c.o
glibPreInstallPhase
installing
install flags: install
[0/1] Installing files.
Installing data/io.elementary.switchboard.screentime-limits.policy to /nix/store/xz0lxir7iglzcw36zmcw3i6ldysbaz1x-switchboard-plug-parental-controls-6.0.0/share/polkit-1/actions/
Installing data/io.elementary.switchboard.parental-controls.appdata.xml to /nix/store/xz0lxir7iglzcw36zmcw3i6ldysbaz1x-switchboard-plug-parental-controls-6.0.0/share/metainfo
Installing src/client/pantheon-parental-controls-client to /nix/store/xz0lxir7iglzcw36zmcw3i6ldysbaz1x-switchboard-plug-parental-controls-6.0.0/libexec
Installing src/daemon/pantheon-parental-controls-daemon to /nix/store/xz0lxir7iglzcw36zmcw3i6ldysbaz1x-switchboard-plug-parental-controls-6.0.0/bin
Installing src/plug/libparental-controls.so to /nix/store/xz0lxir7iglzcw36zmcw3i6ldysbaz1x-switchboard-plug-parental-controls-6.0.0/lib/switchboard/system
Installation failed due to insufficient permissions.
Attempting to use polkit to gain elevated privileges...
pkexec must be setuid root
FAILED: meson-install
/nix/store/452pbrsanrjbz6zj5l7b5b6ywzhi5bw4-meson-0.57.1/bin/meson install --no-rebuild
ninja: build stopped: subcommand failed.
builder for '/nix/store/dvqi8cdzc21gg7qzjz1ccildgg4fjky2-switchboard-plug-parental-controls-6.0.0.drv' failed with exit code 1
error: build of '/nix/store/dvqi8cdzc21gg7qzjz1ccildgg4fjky2-switchboard-plug-parental-controls-6.0.0.drv' failed
```

**Comment by davidak** [comment]
Settings are there, but the backend service seem to be missing.

```
[davidak@gaming:~]$ io.elementary.switchboard 

** (.io.elementary.switchboard-wrapped_:2492): WARNING **: 07:44:29.940: Reader.vala:41: Failed to open file “/etc/security/time.conf”: No such file or directory

** (.io.elementary.switchboard-wrapped_:2492): CRITICAL **: 07:44:29.941: Utils.vala:82: Error calling StartServiceByName for org.pantheon.ParentalControls: Unit pantheon-parental-controls.service not found., using dummy parental controls backend

** (.io.elementary.switchboard-wrapped_:2492): WARNING **: 07:44:29.966: AppsView.vala:232: Unable to get malcontent app filter: App filtering is globally disabled

** (.io.elementary.switchboard-wrapped_:2492): WARNING **: 07:44:41.506: AppsView.vala:189: Failed to set malcontent app filter: GDBus.Error:org.freedesktop.DBus.Error.InvalidArgs: No such interface “com.endlessm.ParentalControls.AppFilter”
```

File: pkgs/desktops/pantheon/apps/switchboard-plugs/parental-controls/default.nix [hunk 1]
```diff
@@ -0,0 +1,69 @@
+{ lib, stdenv
```
Comment (SuperSandro2000) [inline pkgs/desktops/pantheon/apps/switchboard-plugs/parental-controls/default.nix:1]:
```suggestion
{ lib
, stdenv
```

**Comment by bobby285271** [comment]
**TL;DR I haven't manage to get this working yet.**

Just tried this a little bit, but sorry I do not have time and interest to do deep into this yet, but I will post some of my findings.

By checking the build result I guess the package should be added to `services.dbus.packages` and `systemd.packages`. Then you should see `pantheon-parental-controls.service` starts.

After doing these, something like this will happen with probably requires `services.malcontent.enable` to be set.

```
AppsView.vala:232: Unable to get malcontent app filter: App filtering is globally disabled
```

Then this one, I found related file in `linux-pam` package but I am not sure how to deal with that.

```
Writer.vala:88: Failed to open file “/etc/security/time.conf”: No such file or directory
```

And FYI, I think this is hard coded in `src/common/Constants.vala.in`:

```vala
public const string PAM_TIME_CONF_PATH = "/etc/security/time.conf";
```

Then this one:

```
UserConfig.vala:76: Could not find daemon config file: /etc/pantheon-parental-controls/daemon.conf does not exist
```

Again from `src/common/Constants.vala.in` you can find:

```vala
public const string DAEMON_CONF_FILE = "@DAEMON_CONF_DIR@/daemon.conf";
```

Looking at the build result, I think some NixOS options can be created for this but I am not sure if the GUI interface will be still useful since then, or maybe we should put the file in other places?

```plain
# This is main configuration file for pantheon parental controls daemon.
# It is used to restrict users and control whethever they can or cannot
# access a specified service.
#
# a simple example of how a configuration for user "username" would look like:
#  [username]
#  Active=false
#  Targets=/path/to/executable1;/path/to/executable2;
#  BlockUrls=website1.com;website2.com
#  Admin=false
#
# End of example file.
```

Then by checking the source code, I also noticed that in `AccessControlLists.vala`, there is:

```vala
Process.spawn_async (null,
    { "setfacl", "-m", r, target },
    null,
    SpawnFlags.SEARCH_PATH | SpawnFlags.DO_NOT_REAP_CHILD,
    null,
    null);
```

`setfacl` is in `acl` package and I think we probably need to fix the path.

And finally, like other Pantheon components, the locale of this is not set up correctly and you can apply https://github.com/elementary/switchboard-plug-parental-controls/commit/843005b570fc5a661fe6e3f79e3730bc0b1138e8 if you wish, but not that important now I guess.

**Comment by davidak** [comment]
I will continue the work later. Thanks for the input.

**Comment by davidak** [comment]
Just some progress notes:

I got the service, but it was not started.

```
[root@gaming:~]# systemctl status pantheon-parental-controls.service
○ pantheon-parental-controls.service - Pantheon Screen Time & Limits Daemon
     Loaded: loaded (/etc/systemd/system/pantheon-parental-controls.service; linked; vendor preset: enabled)
     Active: inactive (dead)
```

It is only started when the setting is enabled, which makes sense.

Then  i have seen some errors in the log, which i will figure out next.

```
[root@gaming:~]# journalctl -u pantheon-parental-controls.service
-- Journal begins at Tue 2021-08-03 01:49:06 CEST, ends at Sun 2021-10-10 07:09:02 CEST. --
Oct 10 07:08:42 gaming systemd[1]: Starting Pantheon Screen Time & Limits Daemon...
Oct 10 07:08:42 gaming pantheon-parent[7005]: UserConfig.vala:76: Could not find daemon config file: /etc/pantheon-parental-controls/daemon.conf does not exist
Oct 10 07:08:42 gaming pantheon-parent[7005]: UserConfig.vala:167: Failed to create file “etc/pantheon-parental-controls/daemon.conf.IUKEB1”: No such file or directory
Oct 10 07:08:42 gaming systemd[1]: Started Pantheon Screen Time & Limits Daemon.
Oct 10 07:08:42 gaming pantheon-parent[7005]: UserConfig.vala:136: Key file does not have key “BlockUrls” in group “davidak”
Oct 10 07:08:42 gaming pantheon-parent[7005]: UserConfig.vala:120: Key file does not have key “Targets” in group “davidak”
Oct 10 07:08:42 gaming pantheon-parent[7005]: UserConfig.vala:152: Key file does not have key “Admin” in group “davidak”
Oct 10 07:08:55 gaming pantheon-parent[7005]: UserConfig.vala:167: Failed to create file “etc/pantheon-parental-controls/daemon.conf.UTP6A1”: No such file or directory
Oct 10 07:08:55 gaming pantheon-parent[7005]: UserConfig.vala:136: Key file does not have key “BlockUrls” in group “davidak”
Oct 10 07:08:55 gaming pantheon-parent[7005]: Reader.vala:41: Failed to open file “/etc/security/time.conf”: No such file or directory
Oct 10 07:08:55 gaming pantheon-parent[7005]: UserConfig.vala:120: Key file does not have key “Targets” in group “davidak”
Oct 10 07:08:55 gaming pantheon-parent[7005]: UserConfig.vala:152: Key file does not have key “Admin” in group “davidak”
Oct 10 07:09:02 gaming pantheon-parent[7005]: UserConfig.vala:167: Failed to create file “etc/pantheon-parental-controls/daemon.conf.0170A1”: No such file or directory
Oct 10 07:09:02 gaming pantheon-parent[7005]: UserConfig.vala:152: Key file does not have key “Admin” in group “davidak”
Oct 10 07:09:02 gaming pantheon-parent[7005]: Writer.vala:88: Failed to open file “/etc/security/time.conf”: No such file or directory
```

**Comment by davidak** [comment]
> Failed to create file “/etc/pantheon-parental-controls/daemon.conf

Services on NixOS usually have their state in `/var/lib` while `/etc` is declaratively configured using NixOS. So we should move it there.

>Looking at the build result, I think some NixOS options can be created for this but I am not sure if the GUI interface will be still useful

yeah, that would make the GUI obsolete and that is what pantheon users would want to use

**Comment by davidak** [comment]
The `daemon_conf_dir` path is configured here:

https://github.com/elementary/switchboard-plug-parental-controls/blob/49355a30710be1d84e9241fd7b59e1e2d17a38ba/meson.build#L31

So i change `sysconfdir`, which seem to be a meson [built-in option](https://mesonbuild.com/Builtin-options.html).

It's `etc` by default, so i change it to `var/lib`.

Now the error looks like this:

```
Oct 10 12:50:26 gaming systemd[1]: Starting Pantheon Screen Time & Limits Daemon...
Oct 10 12:50:26 gaming pantheon-parent[95630]: UserConfig.vala:76: Could not find daemon config file: /var/lib/pantheon-parental-controls/daemon.conf does not exist
Oct 10 12:50:26 gaming pantheon-parent[95630]: UserConfig.vala:167: Failed to create file “var/lib/pantheon-parental-controls/daemon.conf.UY40A1”: No such file or directory
Oct 10 12:50:26 gaming systemd[1]: Started Pantheon Screen Time & Limits Daemon.
```
I think it's actually the same path, but the output is messed up (https://github.com/elementary/switchboard-plug-parental-controls/issues/154).

**Comment by jtojnar** [comment]
I believe neither Rygel or Geoclue modify the config.

I agree that [`/etc`](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch03s07.html) is the domain of the system administrator and services should not modify it. When a daemon needs to change the data, it should be considered a state and it should probably go to [`/var/lib`](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch05s08.html).

I think this is best done upstream so we should convince Elementary to change `sysconfdir` to `localstatedir`. (We would still need a patch adding `localstatedir_install` but that is something we need to resolve in Meson.)

It is also concerning that `share/dbus-1/system-services/org.pantheon.ParentalControls.service` sets the `User` to `root` (perhaps this is done to be able to edit the file). systemd supports [`StateDirectory`](https://www.freedesktop.org/software/systemd/man/systemd.exec.html#RuntimeDirectory=), which allows it to write to `/var/lib` using unprivileged user. Or does it needs root for connecting to system bus?
